streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -85,27 +85,33 @@ private void initInternal(final ProcessorContext context) {
 
     private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
                                     final InternalProcessorContext context) {
-        final ProcessorRecordContext current = context.recordContext();
-        try {
-            context.setRecordContext(entry.entry().context());
-            if (flushListener != null) {
-                V oldValue = null;
-                if (sendOldValues) {
-                    final byte[] oldBytesValue = underlying.get(entry.key());
-                    oldValue = oldBytesValue == null ? null : serdes.valueFrom(oldBytesValue);
-                }
-                // we rely on underlying store to handle null new value bytes as deletes
-                underlying.put(entry.key(), entry.newValue());
-                flushListener.apply(
-                    serdes.keyFrom(entry.key().get()),
-                    serdes.valueFrom(entry.newValue()),
-                    oldValue,
-                    entry.entry().context().timestamp());
-            } else {
+        if (flushListener != null) {
+            final byte[] newValueBytes = entry.newValue();
+            final byte[] oldValueBytes = underlying.get(entry.key());
+
+            // this is an optimization: if this key did not exist in underlying store and also not in the cache,
+            // we can skip flushing to downstream as well as writing to underlying store
+            if (newValueBytes != null || oldValueBytes != null) {
```

Comment:
Can we apply deMorgan's rule and flip the conditional (with an empty body) here?

It seems easier to understand:
```
if (newValueBytes == null && oldValueBytes == null) {
  // no need to flush or write to underlying store
} else {
  ... the rest of the code
}
```

In other words, the empty "skip" block is more self-documenting than the preceeding comment explaining the algorithm.

Which actually makes me wonder: is there a more general optimization that we can skip desserialization, flush, and write any time the new and old values are identical?

Reply:
I think we cannot generally do this optimization if oldBytes == newBytes, since the timestamp may be updated for `flushListener.apply`.

Reply:
I think an empty then-body would be bad code style.

Reply:
Hey folks, I'm slightly modifying this logic to NOT always blindly read from the underlying store, but do sth. like this:

```
            final byte[] newValueBytes = entry.newValue();
            final byte[] oldValueBytes = newValueBytes != null || sendOldValues ? underlying.get(entry.key()) : null;

            if (newValueBytes != null || oldValueBytes != null) {
                    ....
            }
```

The main motivation is that, for session stores, the likelihood of `newValueBytes == null` could be high while for other two types, the likelihood would be low. As a result, for other two types we would fail the above condition and if `sendOldValues == false` we would not need to read the old bytes for this optimization since it is doomed to not happen.